File: ts/daily-temperatures/daily-temperatures.ts
export function dailyTemperatures(temperatures: number[]): number[] {
  let res: number[] = new Array(temperatures.length).fill(0);
  let decreasingStack: {index: number; value: number}[] = [];
  for (let i = 0; i < temperatures.length; i++) {
    let c = temperatures[i];
    while (decreasingStack.length > 0) {
      let topStack = decreasingStack[decreasingStack.length-1];
      if (topStack.value < c) {
        res[topStack.index] =  i - topStack.index
        decreasingStack.pop()
      } else {
        break
      }
    }
    decreasingStack.push({index: i, value: c})
  }
  return  res;
}

File: py/daily_temperatures/daily_temperatures.py
from typing import List


class Solution:

  def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
    #  [38,37,36,35,39]
    #  [4 ,3 ,2 ,1 ,0]
    # 30 38
    stack = []
    result = [0] * len(temperatures)
    for i, temp in enumerate(temperatures):
      while stack and temp > stack[-1][1]:
        [pos, _] = stack.pop()
        result[pos] = i - pos
      stack.append((i, temp))

    return result

  # def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
  #   #  [38,37,36,35,39]
  #   #  [4 ,3 ,2 ,1 ,0]
  #   # 30 38
  #   stack = []
  #   result = [0] * len(temperatures)
  #   for i, temp in enumerate(temperatures):
  #     while stack and temp > stack[-1]['number']:
  #       number = stack.pop()
  #       result[number['pos']] = number['before']
  #       if stack:
  #         stack[-1]['before'] += number['before']
  #     stack.append({
  #       'number': temp,
  #       'pos': i,
  #       'before': 1,
  #     })
  #
  # return result
